docs/result-type-operator.md

```diff
@@ -0,0 +1,121 @@
+# `result` type function
+
+## Summary
+
+This RFC proposes the addition of a new type function `result`, which can be used to make result types in luau. That are consistent with the idiom established by the [pcall and xpcall global functions](https://luau.org/library#global-functions) in the language currently.
+
+## Motivation
+
+Currently, in order to properly type out a result one would have to write the following:
+
+```luau
+type Mrow = ((meow: string, mrrp: number) -> (true, string)) &
+	((meow: string, mrrp: number) -> (false, nil))
+
+--[[
+	Note: the function is being cast to any, 
+	as this rfc is being written with the assumption the developer is using --!strict mode
+--]]
+local mrow: Mrow = (function(meow: string, mrrp: string)
+	if math.random() > .5 then
+		return true, `{meow} {mrrp}`
+	else
+		return false, nil
+	end
+end) :: any
+```
+
+Due to to that making the developer have to write an overloaded function, and with overloaded functions not consistently working. Most will instead do this:
+
+```luau
+type Mrow = (meow: string, mrrp: number) -> (boolean, string?)
+```
+
+Leading to having the type burden passed onto the developer using the function, as now they have to cast the result if they want to avoid type errors:
+
+```luau
+local success, result = mrow("cat food", ":3")
+
+if success then
+	local new_result: string = result :: any
+	print(`new food of type: {new_result}!`)
+else
+	error("no food :(")
+end	
+```
+
+Some may take a diffrent approach by making their own result type, where they break from the luau idiom:
+
+```luau
+type Result<S, F = nil> = {
+	ok: true,
+	value: S
+} | {
+	ok: false,
+	value: F
+}
+
+local function mrow(meow: string, mrrp: string): Result<string>
+	if math.random() > .5 then
+		return { ok = true, value = `{meow} {mrrp}` }
+	else
+		return { ok = false, value = nil }
+	end
+end
+```
+
+## Design
+
+The functionality of the `result` type function is special, with it being an exception as it'll create a type-pack union. Thus using the result type function will not require the developer to write an overloaded function type.
```

Comment:
Type pack unions do not exist in Luau. So, there's no way for a type pack function to produce a type pack union.

Reply:
Yes, but result would be special cased. As its the only use for typepack unions.

Reply:
There's not a way for us to do that right now. Type functions and type pack functions have to return types or type packs, and type packs are not unions. Doesn't mean we can't do this at all, but it does mean that this is essentially the same amount of work as proposing general support for type pack unions.